internals/cli/masker/writer.go

```diff
@@ -0,0 +1,204 @@
+	for offset := 1; offset <= m.currentIndex; offset++ {
+		ok := true
+		for i := 0; i < m.currentIndex-offset; i++ {
+			if m.sequence[i] != m.sequence[i+offset] {
+				ok = false
+				break
+			}
+		}
+		if ok {
+			m.currentIndex -= offset
+			return
+		}
+	}
+}
+
+// InProgress returns whether this sequenceMatcher is currently partially matching.
+//
+// For example, if the sequence is "foobar" and the registered input is "foob", InProgress() returns true.
+func (m *sequenceMatcher) InProgress() bool {
+	return m.currentIndex > 0
+}
+
+// Reset forgets the current match.
+func (m *sequenceMatcher) Reset() {
+	m.currentIndex = 0
+}
+
+// maskByte represents a byte and whether the byte should be masked or not.
+type maskByte struct {
+	byte
+	masked bool
+}
+
+// MaskedWriter wraps an io.Writer and masks all matches found by Matchers with MaskString.
+// If no write is made for Timeout on the io.Writer, any matches in progress are reset
+// and the buffer is flushed. This is to ensure that the writer does not hang on partial matches.
+type MaskedWriter struct {
+	w          io.Writer
+	MaskString string
+	Matchers   []Matcher
+	Timeout    time.Duration
+
+	buf    []maskByte
+	lock   *sync.Mutex
+	output chan []maskByte
+	err    error
+	nIn    int64
+	nOut   int64
+}
+
+func NewMaskedWriter(w io.Writer, masks [][]byte, maskString string, timeout time.Duration) *MaskedWriter {
+	var lock sync.Mutex
+	matchers := make([]Matcher, len(masks))
+	for i, mask := range masks {
+		matchers[i] = &sequenceMatcher{
+			sequence: mask,
+		}
+	}
+	return &MaskedWriter{
+		w:          w,
+		MaskString: maskString,
+		Matchers:   matchers,
+		Timeout:    timeout,
+		lock:       &lock,
+		output:     make(chan []maskByte, 1),
+	}
+}
+
+// Write implements implements to Write from io.Writer
+// It is responsible for finding any matches to mask and mark the appropriate bytes as masked.
+// This function never returns an error. These can instead be caught with Flush().
+func (mw *MaskedWriter) Write(p []byte) (n int, err error) {
+	for _, b := range p {
+		matchInProgress := false
+
+		mw.lock.Lock()
+		mw.buf = append(mw.buf, maskByte{byte: b})
+
+		for _, matcher := range mw.Matchers {
+			maskLen := matcher.Read(b)
+			for i := 0; i < maskLen; i++ {
+				mw.buf[len(mw.buf)-1-i].masked = true
+			}
+			matchInProgress = matchInProgress || matcher.InProgress()
+		}
+
+		if !matchInProgress {
+			mw.flushBuffer()
+		} else {
+			mw.output <- []maskByte{}
+		}
+
+		mw.lock.Unlock()
+	}
+
+	mw.nIn += int64(len(p))
+
+	return len(p), nil
+}
+
+func (mw *MaskedWriter) flushBuffer() {
+	tmp := make([]maskByte, len(mw.buf))
+	copy(tmp, mw.buf)
+	mw.output <- tmp
+	mw.buf = mw.buf[:0]
+}
+
+// Run writes any data processed data from the output channel to the underlying io.Writer.
+// If no new data is received on the output channel for Timeout, the output buffer is forced flushed
+// and all ongoing matches are reset.
+//
+// This should be run in a separate goroutine.
```

Comment:
If this should always be run in a separate goroutine, does is make sense to start that goroutine in this function?

Reply:
As far as I know, idiomatic Go prescribes that you should leave this to the caller. Let me look that up.

Reply:
https://talks.golang.org/2013/bestpractices.slide#25
https://stackoverflow.com/questions/28506013/idiomatic-golang-goroutines
https://hackmongo.com/page/golang-antipatterns/#hidden-goroutine

Reply:
Let's keep it as is :+1: